subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptHandler.java

```diff
@@ -99,9 +100,12 @@ public ClassPath getScriptClassPath() {
     public ClassPath getInstrumentedScriptClassPath() {
         if (resolvedClasspath == null) {
             if (classpathConfiguration != null) {
-                resolvedClasspath = buildLogicBuilder.resolveClassPath(classpathConfiguration, dependencyHandler, configContainer);
-                if (!getBoolean(DISABLE_RESET_CONFIGURATION_SYSTEM_PROPERTY)) {
-                    ((ResettableConfiguration) classpathConfiguration).resetResolutionState();
+
+                Factory<ClassPath> classPathFactory = () -> buildLogicBuilder.resolveClassPath(classpathConfiguration, dependencyHandler, configContainer);
+                if (getBoolean(DISABLE_RESET_CONFIGURATION_SYSTEM_PROPERTY)) {
+                    resolvedClasspath = classPathFactory.create();
+                } else {
+                    resolvedClasspath = ((ResettableConfiguration) classpathConfiguration).withResetResolutionState(classPathFactory);
```

Comment:
Do I understand correctly that this change keeps the dependency substitution data alive in all cases, not just when `buildEnvironment` is used?

Reply:
Yes, that's the case.  We keep the substitution rules for the buildscript classpath configuration only, in case it is re-resolved by something (which may be `buildEnvironment` but could be anything else that tries to resolve the classpath configuration for some reason).  However, once something else resolves the classpath configuration, we treat it like any other configuration at that point and cache the resolution results, discard the substitution rules, etc.